text/0000-pack-cache-options.md

```diff
@@ -0,0 +1,222 @@
+
+# Motivation
+[motivation]: #motivation
+
+The platform spec currently provides two formats for caching build layers. These formats are a directory or an image.
+
+Additionally, there is an option to specify a _`launch-cache`_ directory to cache launch layers.
+
+> NOTE: Currently _`launch-cache` flag_ is only used by the current implementation of the lifecycle when _`-daemon` flag_ is provided.
+
+Pack, as a consumer of the lifecycle using the platform spec, uses these options in the following ways:
+
+* By default, `pack build` would create two named volumes
+
+<!-- - Why should we do this?
+- What use cases does it support?
+- What is the expected outcome? -->
+
+##### Use Cases
+
+* As a Pack user, I would like to build using a docker daemon while storing _`build` cache_ as an image (aka _`cache-image`_) in a registry. ([buildpacks/pack#1024](https://github.com/buildpacks/pack/issues/1024#issuecomment-762901676))
+* As a Pack user using `pack` on BitBucket Pipelines, I would like to be able to store the _`build` cache_ and _`launch` cache_ on the file system. ([bitbucket discussion](https://community.atlassian.com/t5/Bitbucket-discussions/Using-CNCF-Buildpacks-on-Bitbucket-Pipelines/td-p/1516690))
+* As a Pack user, I would like to be able to specify the name of the volumes used by pack for caching. ([buildpacks/pack#665](https://github.com/buildpacks/pack/issues/665))
+
+# What it is
+[what-it-is]: #what-it-is
+
+<!-- This provides a high level overview of the feature.
+
+- Define any new terminology.
+- Define the target persona: buildpack author, buildpack user, platform operator, platform implementor, and/or project contributor.
+- Explaining the feature largely in terms of examples.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing users and new users. -->
+
+A new variadic flag that enables configuring different cache types geared towards app developers or operators that are using pack in very specific (and sometimes restrictive) ways that require them to configure caching in a more advanced fashion.
+
+The options available via this new flag would have sensible defaults based on the current behaiviour with the exception of the `--cache-image` which would be deprecated and later removed to reduce confusion. Overall UX would remain the same while allowing for greater flexibility.
+
+An additional benefit of this format is that it can later be expanded on if more options for a specific format were deamed necessary.
+
+```
+pack build --cache 'type=<type>;format=<format>;<additional-options>' --cache ...
+```
+
+### `cache` flag options
+* _**type**_ - The type of cache to configure [`build`, `launch`] (default: `build`).
+* _**format**_ - The format in which to store that cache [`dir`, `image`, `volume`] (default: `volume`). 
```

Comment:
I am hesitant to expose the launch cache concept to users. It is a hack necessary to produce adequate performance when exporting to a docker daemon. Instead f exposing this concept to users I would rather drop it entirely if we stop supporting the daemon as an export target (as we have previously discussed).

Making it directly configurable only opens up potential confusion imho, for example if we used this cache during a registry build it would actually decrease performance. Also, neither layer contents or layer metadata are restored from this cache, making it confusing to buildpack authors concerned with the layer reuse lifecycle.

Reply:
Do we have existing use cases for configuring the launch cache? Could we solve this problem another way?

Reply:
Think [the linked ticket](https://community.atlassian.com/t5/Bitbucket-discussions/Using-CNCF-Buildpacks-on-Bitbucket-Pipelines/td-p/1516690) would require the launch cache to be configurable. We could potentially work around this by making the following changes in pack:
- Explicitly create a launch cache volume before running any of the lifecycle phases.
  - If this errors out because of permissions just ignore it and continue on without a launch cache
  - If it works yay mount it into the lifecycle containers!
- Force people to use image caches in these environments (this is already an ask [here](https://github.com/buildpacks/pack/issues/1024))

Reply:
As @dwillist mentions, we _need_ to expose users to the launch cache in order to be able to configure it for very specific use cases. Another example just came up in pack-orb (https://github.com/buildpacks/pack-orb/issues/25#issuecomment-824086257).

> The reason for the current requirement appears to be on trying to cache the docker volumes created. Once we have a way to configure cache volumes via pack, we should be able to remove that dependency.

I'm on the boat for removing it and IMO that is one of the benefits of this proposal. In the future, the `type` option would likely not exist.